src/queue/simple.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A simple task queue.
+//!
+//! This task queue requires [`TaskCell`] to contain [`SimpleQueueExtras`]
+//! as its extras. The instant when the task cell is pushed into the queue
+//! is recorded in the extras.
+
+use super::{LocalQueue, Pop, TaskCell, TaskInjector};
+
+use crossbeam_deque::{Injector, Steal, Stealer, Worker};
+use std::iter;
+use std::sync::Arc;
+use std::time::Instant;
+
+/// The injector of a simple task queue.
+pub struct SimpleQueueInjector<T>(Arc<Injector<T>>);
+
+impl<T: TaskCell> Clone for SimpleQueueInjector<T> {
+    fn clone(&self) -> Self {
+        SimpleQueueInjector(self.0.clone())
+    }
+}
+
+/// The extras for the task cells pushed into a simple task queue.
+///
+/// [`Default::default`] can be used to create a [`SimpleQueueExtras`] for
+/// a task cell.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct SimpleQueueExtras {
+    /// The instant when the task cell is pushed to the queue.
+    schedule_time: Option<Instant>,
+}
+
+fn set_schedule_time<T>(task_cell: &mut T)
+where
+    T: TaskCell<Extras = SimpleQueueExtras>,
+{
+    task_cell.mut_extras().schedule_time = Some(Instant::now());
+}
+
+impl<T> TaskInjector for SimpleQueueInjector<T>
+where
+    T: TaskCell<Extras = SimpleQueueExtras>,
+{
+    type TaskCell = T;
+
+    /// Pushes the task cell to the queue. The schedule time in the extras is
+    /// assigned to be now.
+    fn push(&self, mut task_cell: Self::TaskCell) {
+        set_schedule_time(&mut task_cell);
+        self.0.push(task_cell);
+    }
+}
+
+/// The local queue of a simple task queue.
+pub struct SimpleQueueLocal<T> {
+    local_queue: Worker<T>,
+    injector: Arc<Injector<T>>,
+    stealers: Vec<Stealer<T>>,
+    self_index: usize,
+}
+
+impl<T> LocalQueue for SimpleQueueLocal<T>
+where
+    T: TaskCell<Extras = SimpleQueueExtras>,
+{
+    type TaskCell = T;
+
+    fn push(&mut self, mut task_cell: Self::TaskCell) {
+        set_schedule_time(&mut task_cell);
+        self.local_queue.push(task_cell);
+    }
+
+    fn pop(&mut self) -> Option<Pop<Self::TaskCell>> {
+        fn into_pop<T>(mut t: T, from_local: bool) -> Pop<T>
+        where
+            T: TaskCell<Extras = SimpleQueueExtras>,
+        {
+            let schedule_time = t.mut_extras().schedule_time.unwrap();
+            Pop {
+                task_cell: t,
+                schedule_time,
+                from_local,
+            }
+        }
+
+        if let Some(t) = self.local_queue.pop() {
+            return Some(into_pop(t, true));
+        }
+        let mut need_retry;
+        loop {
+            need_retry = false;
+            match self.injector.steal_batch_and_pop(&self.local_queue) {
+                Steal::Success(t) => return Some(into_pop(t, false)),
+                Steal::Retry => need_retry = true,
+                _ => {}
+            }
+            for (i, stealer) in self.stealers.iter().enumerate() {
```

Comment:
Technically it can lead to unbalance because tasks of queues at smaller index will always get more chance to be stolen.

Reply:
OK. I can use a rng to solve it.